backend/eventloop/shared_resource_loop/sharedresourceloop_test.go

```diff
@@ -0,0 +1,345 @@
+				Build()
+
+			k8sClient := &sharedutil.ProxyClient{
+				InnerClient: k8sClientOuter,
+				Informer:    &informer,
+			}
+
+			sharedResourceEventLoop := &SharedResourceEventLoop{inputChannel: make(chan sharedResourceLoopMessage)}
+
+			go internalSharedResourceEventLoop(sharedResourceEventLoop.inputChannel)
+
+			// At first assuming there are no existing users, hence creating new.
+			usrOld,
+				err,
+				isNewUser := sharedResourceEventLoop.GetOrCreateClusterUserByNamespaceUID(ctx, k8sClient, *workspace)
+
+			// Check error in Nil and user is not Nil
+			Expect(err).To(BeNil())
+			Expect(usrOld).NotTo(BeNil())
+
+			// Check new user is created
+			Expect(isNewUser).To(BeTrue())
+
+			// User is created in previous call, then same user should be returned instead of creating new.
+			usrNew,
+				err,
+				isNewUser := sharedResourceEventLoop.GetOrCreateClusterUserByNamespaceUID(ctx, k8sClient, *workspace)
+
+			// Check error in Nil and user is not Nil
+			Expect(err).To(BeNil())
+			Expect(usrNew).NotTo(BeNil())
+
+			// Check user is not recreated
+			Expect(isNewUser).To(BeFalse())
+
+			// Check old and new users are same
+			Expect(usrOld).To(Equal(usrNew))
+
+			// To be used by AfterEach to clean up the resources created by test
+			resourcesToBeDeleted = testResources{Clusteruser_id: usrNew.Clusteruser_id}
+		})
+
+		It("Should create or fetch resources.", func() {
+			// Create a fake K8s client
+			ctx := context.Background()
+
+			scheme,
+				argocdNamespace,
+				kubesystemNamespace,
+				workspace, _ := eventlooptypes.GenericTestSetup()
+
+			gitopsDepl := &managedgitopsv1alpha1.GitOpsDeployment{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "my-gitops-depl",
+					Namespace: workspace.Name,
+					UID:       uuid.NewUUID(),
+				},
+			}
+
+			informer := sharedutil.ListEventReceiver{}
+
+			k8sClientOuter := fake.NewClientBuilder().
+				WithScheme(scheme).
+				WithObjects(gitopsDepl, workspace, argocdNamespace, kubesystemNamespace).
+				Build()
+
+			k8sClient := &sharedutil.ProxyClient{
+				InnerClient: k8sClientOuter,
+				Informer:    &informer,
+			}
+
+			sharedResourceEventLoop := &SharedResourceEventLoop{inputChannel: make(chan sharedResourceLoopMessage)}
+
+			go internalSharedResourceEventLoop(sharedResourceEventLoop.inputChannel)
+
+			// At first assuming there are no existing resources, hence creating new.
+			clusterUserOld,
+				isNewUser,
+				managedEnvOld,
+				isNewManagedEnv,
+				gitopsEngineInstanceOld,
+				isNewInstance,
+				clusterAccessOld,
+				isNewClusterAccess,
+				gitopsEngineClusterOld,
+				err := sharedResourceEventLoop.GetOrCreateSharedResources(ctx, k8sClient, *workspace)
+
+			// Check if result is not Nil and error is Nil
+			Expect(err).To(BeNil())
+			Expect(clusterUserOld).NotTo(BeNil())
+			Expect(managedEnvOld).NotTo(BeNil())
+			Expect(gitopsEngineInstanceOld).NotTo(BeNil())
+			Expect(clusterAccessOld).NotTo(BeNil())
```

Comment:
Do we need to check 'gitopsEngineClusterOld', too? Noticed it was missing

Reply:
It is not used anywhere for test, but it it is created prior to GitOps Engine Instance, so it needs to be deleted while cleaning up resources, that's why it is returned back to test and then sent to `DeferCleanup` function.
https://github.com/redhat-appstudio/managed-gitops/blob/8c72ce9bc207c389e260a2808514083accbe42cf/backend-shared/config/db/util/utils.go#L125